opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,676 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "angles/angles.h"
+#include "opennav_docking_core/docking_exceptions.hpp"
+#include "opennav_following/following_server.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include "tf2/utils.hpp"
+
+using namespace std::chrono_literals;
+using rcl_interfaces::msg::ParameterType;
+using std::placeholders::_1;
+
+namespace opennav_following
+{
+
+FollowingServer::FollowingServer(const rclcpp::NodeOptions & options)
+: nav2::LifecycleNode("following_server", "", options)
+{
+  RCLCPP_INFO(get_logger(), "Creating %s", get_name());
+
+  declare_parameter("controller_frequency", 50.0);
+  declare_parameter("detection_timeout", 2.0);
+  declare_parameter("rotate_to_object_timeout", 10.0);
+  declare_parameter("static_object_timeout", -1.0);
+  declare_parameter("linear_tolerance", 0.15);
+  declare_parameter("angular_tolerance", 0.15);
+  declare_parameter("max_retries", 3);
+  declare_parameter("base_frame", "base_link");
+  declare_parameter("fixed_frame", "odom");
+  declare_parameter("allow_backward", false);
+  declare_parameter("filter_coef", 0.1);
+  declare_parameter("desired_distance", 1.0);
+  declare_parameter("skip_orientation", true);
+  declare_parameter("search_by_rotating", false);
+  declare_parameter("odom_topic", "odom");
+  declare_parameter("transform_tolerance", 0.1);
+}
+
+nav2::CallbackReturn
+FollowingServer::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring %s", get_name());
+  auto node = shared_from_this();
+
+  get_parameter("controller_frequency", controller_frequency_);
+  get_parameter("detection_timeout", detection_timeout_);
+  get_parameter("rotate_to_object_timeout", rotate_to_object_timeout_);
+  get_parameter("static_object_timeout", static_object_timeout_);
+  get_parameter("linear_tolerance", linear_tolerance_);
+  get_parameter("angular_tolerance", angular_tolerance_);
+  get_parameter("max_retries", max_retries_);
+  get_parameter("base_frame", base_frame_);
+  get_parameter("fixed_frame", fixed_frame_);
+  get_parameter("allow_backward", allow_backward_);
+  get_parameter("desired_distance", desired_distance_);
+  get_parameter("skip_orientation", skip_orientation_);
+  get_parameter("search_by_rotating", search_by_rotating_);
+  get_parameter("transform_tolerance", transform_tolerance_);
+  RCLCPP_INFO(get_logger(), "Controller frequency set to %.4fHz", controller_frequency_);
+
+  vel_publisher_ = std::make_unique<nav2_util::TwistPublisher>(node, "cmd_vel");
+  tf2_buffer_ = std::make_shared<tf2_ros::Buffer>(node->get_clock());
+
+  // Create odom subscriber for backward blind docking
+  std::string odom_topic;
+  get_parameter("odom_topic", odom_topic);
+  odom_sub_ = std::make_unique<nav_2d_utils::OdomSubscriber>(node, odom_topic);
```

Comment:
See https://github.com/ros-navigation/navigation2/pull/5414: We should update to use the OdomSmoother in Nav2 utils. We're deprecating use of nav_2d_utils